### PR TITLE
Not serialize compile arguments which have no S11N<T> template specialization

### PR DESCRIPTION
**Pull request by AsyaPronina**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [X] I agree to contribute to the project under Apache 2 License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [X] The PR is proposed to proper branch
- [X] There is reference to original bug report and related work
- [X] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [X] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom,Custom Win,Custom Mac
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

build_image:Custom=centos:7
buildworker:Custom=linux-1
build_gapi_standalone:Custom=ade-0.1.1f

Xbuild_image:Custom=ubuntu-openvino-2020.3.0:16.04
build_image:Custom Win=openvino-2020.3.0
build_image:Custom Mac=openvino-2020.3.0

Xtest_modules:Custom=gapi
test_modules:Custom Win=gapi
test_modules:Custom Mac=gapi

Xbuildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
Xtest_opencl:Custom=OFF
Xtest_bigdata:Custom=1
Xtest_filter:Custom=*
```

### CODE REVIEW

**Comment by AsyaPronina** [comment]
Cases where user passes more than 1 variable of certain type as compile arguments to ``cv::gapi::serialize(GCompileArgs)`` are not handled.

File: modules/gapi/src/backends/common/serialization.cpp [hunk 1]
```diff
@@ -338,8 +338,13 @@ IIStream& operator>> (IIStream& is,       cv::gapi::wip::draw::Line &l) {
 
 IOStream& operator<< (IOStream& os, const cv::GCompileArg& arg)
 {
-    os << arg.tag;
-    arg.serialize(os);
+    ByteMemoryOutStream tmpS;
+    tmpS << arg.tag;
```
Comment (dmatveev) [inline modules/gapi/src/backends/common/serialization.cpp:342]: Tag could be written to a primary stream.

File: modules/gapi/include/opencv2/gapi/s11n.hpp [hunk 2]
```diff
@@ -265,23 +265,25 @@ void getRunArgByIdx (IIStream& is, cv::util::variant<Ts...> &v, uint32_t idx) {
...
-struct deserialize_arg<std::tuple<T, Types...>> {
-static GCompileArg exec(cv::gapi::s11n::IIStream& is, const std::string& tag) {
+struct try_deserialize_comparg<std::tuple<T, Types...>> {
+static cv::util::optional<GCompileArg> exec(cv::gapi::s11n::IIStream& is, const std::string& tag) {
     if (tag == cv::detail::CompileArgTag<T>::tag()) {
-        return GCompileArg {
-            cv::gapi::s11n::detail::S11N<T>::deserialize(is)
-        };
+        static_assert(cv::gapi::s11n::detail::has_S11N_spec<T>::value,
+            "cv::gapi::deserialize<GCompileArgs, Types...> waits Types to have S11N specializations"
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/s11n.hpp:281]: `waits` -> `expects`

File: modules/gapi/include/opencv2/gapi/s11n.hpp [hunk 3]
```diff
@@ -303,17 +305,36 @@ static GRunArg exec(cv::gapi::s11n::IIStream& is, uint32_t idx) {
...
 
 template<typename... Types>
-cv::GCompileArgs getCompileArgs(const std::vector<char> &p) {
-    std::unique_ptr<cv::gapi::s11n::IIStream> pIs = cv::gapi::s11n::detail::getInStream(p);
+cv::util::optional<GCompileArg> tryDeserializeCompArg(const std::vector<char>& sArg) {
+    std::unique_ptr<cv::gapi::s11n::IIStream> pIs = cv::gapi::s11n::detail::getInStream(sArg);
     cv::gapi::s11n::IIStream& is = *pIs;
+
+    std::string tag;
+    is >> tag;
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/s11n.hpp:313]: Here you read `tag` directly from the stream, but in the serialization code, you write tag to a temporary stream and then serialize its vector.

In theory, this code should work incorrectly. Please double-check.

https://github.com/opencv/opencv/pull/18651/files#diff-7b0234dbbb6f5f5bae56c90408ad90fe57c7d3070a1ed19af9ee4c909289685fR342

File: modules/gapi/test/s11n/gapi_s11n_tests.cpp [hunk 4]
```diff
@@ -596,4 +674,151 @@ TEST_F(S11N_Basic, Test_Custom_CompileArg) {
...
+    EXPECT_EQ(myCustomVar, cv::gapi::getCompileArg<MyCustomType>(dArgs3).value());
+    EXPECT_TRUE(cv::gapi::getCompileArg<EmptyCustomType>(dArgs3).has_value());
+    EXPECT_EQ(simpleCustomVar, cv::gapi::getCompileArg<SimpleCustomType>(dArgs3).value());
+
+    //----{ emptyCustomVar }----
+    std::vector<char> sArgs4 = cv::gapi::serialize(cv::compile_args(emptyCustomVar));
+    GCompileArgs dArgsEmptyVar4 = cv::gapi::deserialize<GCompileArgs, EmptyCustomType>(sArgs4);
+    EXPECT_EQ(1u, dArgsEmptyVar4.size());
+    EXPECT_TRUE(cv::gapi::getCompileArg<EmptyCustomType>(dArgsEmptyVar4).has_value());
+}
```
Comment (dmatveev) [inline modules/gapi/test/s11n/gapi_s11n_tests.cpp:822]: There must be a test failing due to inconsistency in S11N for custom compile arg tags.

File: modules/gapi/include/opencv2/gapi/s11n.hpp (hunk 3 above)
Comment (AsyaPronina) [reply modules/gapi/include/opencv2/gapi/s11n.hpp:313]: Here I have an GCompileArg as serialized vector of char, which were deserialized from the main stream and this vector of char should contain that that were deserialized in temporary stream

Comment (dmatveev) [reply modules/gapi/include/opencv2/gapi/s11n.hpp:313]: Please simplify, the code shouldn't cause the confusion like mine

File: modules/gapi/src/backends/common/serialization.cpp (hunk 1 above)
Comment (AsyaPronina) [reply modules/gapi/src/backends/common/serialization.cpp]: Thanks! Reworked this part!

File: modules/gapi/include/opencv2/gapi/s11n.hpp (hunk 2 above)
Comment (AsyaPronina) [reply modules/gapi/include/opencv2/gapi/s11n.hpp]: Fixed

File: modules/gapi/include/opencv2/gapi/s11n.hpp (hunk 3 above)
Comment (AsyaPronina) [reply modules/gapi/include/opencv2/gapi/s11n.hpp]: Simplified, thanks!

File: modules/gapi/test/s11n/gapi_s11n_tests.cpp (hunk 4 above)
Comment (AsyaPronina) [reply modules/gapi/test/s11n/gapi_s11n_tests.cpp:822]: Discussed

File: modules/gapi/include/opencv2/gapi/s11n.hpp [hunk 5]
```diff
@@ -303,17 +305,35 @@ static GRunArg exec(cv::gapi::s11n::IIStream& is, uint32_t idx) {
...
     is >> sz;
     for (uint32_t i = 0; i < sz; ++i) {
         std::string tag;
         is >> tag;
-        args.push_back(cv::gapi::detail::deserialize_arg<std::tuple<Types...>>::exec(is, tag));
+
+        std::vector<char> sArg;
+        is >> sArg;
+
+        cv::util::optional<GCompileArg> dArg =
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/s11n.hpp:330]: This whole thing can be further simplified, but let's leave it for now.
Just put a task in our tracker to implement GCompileArgs serialization in a more generic two-stage way (based on vectors).

**Review by dmatveev — APPROVED** [review]

**Review by smirnov-alexey — APPROVED** [review]
Overall looks good

**Pull request merged**